[Query Rules] Add documentation for rule_query (#97667)
elastic/elasticsearch: docs/reference/search/search-your-data/search-using-query-rules.asciidoc

Reviewer: For this doc, I think the technical details are good, but I'm working pretty hard to piece together the example from the perspective of the business use case.

How I understand the example: If a user searches for "pugs", pin two specific documents to the top of the results. Assuming I have that right, I suggest stating that explicitly in the introduction. It will make the rest easier to follow.

Author: That's correct. At first the only supported rule is a pinned query but we may add more in subsequent releases. I'll make sure to make it clearer - thank you!

Author: Updated, hopefully this is more clear w.r.t. value and whats happening.

Reviewer: Yes, nice improvements!

[[search-using-query-rules]]
== Searching with query rules
++++
<titleabbrev>Searching with query rules</titleabbrev>
++++

[[query-rules]]

preview::[]

_Query rules_ are rules that are defined inside {es} and are applied only to queries that match these rules.
This allows more control over promoted results for specific queries only. For example, you may want to promote results only when users are searching for specific items.

Rules are defined using the <<query-rules-apis,Query rules API>>.

> When defining a rule, we take into consideration the following:

Reviewer: I'm not sure who "we" is in this sentence and list. Author and reader? Elasticsearch?

If author/reader, I'd revise to avoid "we". For example: "When defining a rule, consider the following:".

If Elasticsearch, replace "we" with "Elasticsearch".

> * The type of rule we want to apply.

Reviewer: Could add quick examples here of what is meant by type, criteria, and actions. Otherwise, they don't mean anything until you start parsing the code examples below.

I was also initially confused because the rule has a type and the criteria has a type.

* The criteria for which this rule will match.
* The actions to take when the rule matches a query.

Current supported rule types:

* `pinned` query rules will re-write the query into a <<query-dsl-pinned-query, pinned query>>, promoting specific results matching the query rule at the top of the returned result set.

[discrete]
[[add-query-rules]]
=== Add query rules

You can add query rules using the <<put-query-ruleset>> call. This adds a ruleset containing one or more query rules that will be applied to queries that match their specified criteria.

The following command will create a query ruleset called `my-ruleset` with two pinned document rules:

* The first rule will generate a <<query-dsl-pinned-query>> pinning the <<mapping-id-field,`_id`>>s `id1` and `id2` when the `query_string` matches the value `pugs`.
* The second rule will generate a <<query-dsl-pinned-query>> pinning the <<mapping-id-field, `_id`>> of `id3` specifically from the `my-index-000001` index and `id4` from the `my-index-000002` index when the `query_string` matches the value `beagles`.

[source,console]
----
PUT /_query_rules/my-ruleset
{

>   "rules": [

Author: > Nit: I suggest we be consistent with the indentation, at least within the same page (all code blocks with 2 spaces OR 4 spaces).

@demjened I was following the model here for other query DSL docs, such as [pinned query](https://github.com/elastic/elasticsearch/blob/main/docs/reference/query-dsl/pinned-query.asciidoc).

>   "rules": [

Reviewer 2: That page looks consistent with 2 spaces, which is good. I'm referring to the code blocks in these pages:
- in `rule_query.asciidoc` the `PUT _query_rules/my-ruleset` example uses 4 spaces, then right underneath `GET /_search` uses 2
- In `search-using-query-riles.asciidoc` `PUT /_query_rules/my-ruleset` uses 2 spaces, the response has 4 and the ``GET /my-index-000001/_search` also has 4

        {
          "type": "exact",
          "metadata": "query_string",
          "value": "beagles"
        }
      ],
      "actions": {
        "docs": [
          {
            "_index": "my-index-000001",
            "_id": "id3"
          },
          {
            "_index": "my-index-000002",
            "_id": "id4"
          }
        ]
      }
    }
  ]
}
----
// TEST[skip:TBD]

The API response returns a results of `created` or `updated` depending on whether this was a new or edited ruleset.

[source,console-result]
----
{
    "result": "created"
}
----
// TEST[skip:TBD]

You can use the <<get-query-ruleset>> call to retrieve the ruleset you just created,
the <<list-query-rulesets>> call to retrieve a summary of all query rulesets,
and the <<delete-query-ruleset>> call to delete a query ruleset.

[discrete]
[[rule-query-search]]
=== Perform a rule query

Once you have defined a query ruleset, you can search this ruleset using the <<query-dsl-rule-query>> query.
An example query for the `my-ruleset` defined above is:

[source,console]
----
GET /my-index-000001/_search
{
    "query": {
        "rule_query": {
            "organic": {
                "query_string": {

>                   "query": "pugs"

Reviewer 2: Nit: indentation fix (4 spaces).
```suggestion
                    "query": "pugs"
```

[[search-using-query-rules]]
== Searching with query rules
++++
<titleabbrev>Searching with query rules</titleabbrev>
++++

[[query-rules]]

preview::[]

> _Query rules_ are rules that are defined inside {es} that allow customization of search results for queries that match specified criteria metadata.

Author: Nice simplification, thank you!

> This allows for more control over promoted results, ensuring that specified documents always match defined criteria and are returned at the top of the result list.

Reviewer 3: I think we should keep open the goal of query rules. Maybe we could move this to the pinned query rule type, or add a paragraph explaining that right now this is the only use for query rules...

Author: I moved this here based on earlier feedback, to make it less confusing that _right now_ query rules are for pinned documents only. I added a small clarification that should help, though, WDYT?

Metadata is defined in the query rule, and is matched against the query criteria.

> Metadata is a string key, and can be anything from a user-entered query string, personalized metadata about users (e.g. country, language, etc). or anything else defined in the rules.

Reviewer 3: I find metadata definition a bit confusing here. I was thinking on how to rephrase this:

```suggestion
Query rules use metadata to match a query.
Metadata is provided as part of the `rule_query` as an object. 
It can be anything from a user-entered query string, personalized metadata about users (e.g. country, language, etc).

Query rules define a metadata key that will be used to match the metadata provided in the `rule_query` with the criteria specified in the rule.

When a query rule match the `rule_query` metadata according to its defined criteria, the query rule action is applied to the underlying `organic_query`
```

... maybe adding some small examples at the end of each paragraph will make sense as well to guide the user step by step on what query rules are, what metadata is, and how they match.

Author: Good call out, I can leave the personalization stuff for a blog post 👍

Author: OK, I've incorporated your suggestions, LMK what you think.

Rules are defined using the <<query-rules-apis,Query rules API>> and searched using the <<query-dsl-rule-query,rule query>>.

When defining a rule, consider the following:

> * The type of rule we want to apply:

Reviewer 3: ```suggestion
* The type of rule we want to apply. For the moment, there is a single rule type:
```

Author: Good catch!

+
--
* `pinned` will re-write the query into a <<query-dsl-pinned-query, pinned query>>, pinning specified results matching the query rule at the top of the returned result set.
--

> * The criteria for which this rule will match. Criteria is defined as `type`, `metadata`, and `values`. The allowed types are:

Reviewer 3: Would adding section headers help to divide the rule structure, and provide helpful navigation?

Author: Added, thank you for the suggestion!

+
--

> * `exact` will match when the rule metadata matches the specified value exactly.

Reviewer 3: Would a table be a better choice for having the type, description, and an example of match?

For example:
| prefix | matches when the rule metadata starts with the specified value | "pref" will match "prefix" |

* `fuzzy` will match when the rule metadata matches the specified value within an allowed {wikipedia}/Levenshtein_distance[Levenshtein edit distance].
* `prefix` will match when the rule metadata starts with the specified value.
* `suffix` will match when the rule metadata ends with the specified value.
* `contains` will match when the rule metadata contains the specified value.
* `lt` will match when the rule metadata is less than the specified value.
* `lte` will match when the rule metadata is less than or equal to the specified value.
* `gt` will match when the rule metadata is greater than the specified value.
* `gte` will match when the rule metadata is greater than or equal to the specified value.
* `always` will always match for all rule queries.
--
* The actions to take when the rule matches a query:
+
--

> * `ids` will pin the specified <<mapping-id-field,`_id`>>s.

Reviewer 3: ```suggestion
* `ids` will pin the specified <<mapping-id-field,`_id`>>s from a single index.
```

Author: _Technically_ if we have the same ID in two indices then both will be pinned.

* `docs` will pin the specified documents in the specified indices.
--

> 

Reviewer 3: ```suggestion

Use `ids` when searching over a single index, and `docs` when searching over multiple indices. See <<query-dsl-pinned-query,pinned query>> for details.

```

Author: Good to add for now, may change in the future.

[discrete]
[[add-query-rules]]
=== Add query rules

You can add query rules using the <<put-query-ruleset>> call. This adds a ruleset containing one or more query rules that will be applied to queries that match their specified criteria.

The following command will create a query ruleset called `my-ruleset` with two pinned document rules:

> * The first rule will generate a <<query-dsl-pinned-query>> pinning the <<mapping-id-field,`_id`>>s `id1` and `id2` when the `user.query` matches the value `pugs`.

Reviewer 3: ```suggestion
* The first rule will generate a <<query-dsl-pinned-query>> pinning the <<mapping-id-field,`_id`>>s `id1` and `id2` when the `user.query` metadata value matches `pugs`.
```

> * The second rule will generate a <<query-dsl-pinned-query>> pinning the <<mapping-id-field, `_id`>> of `id3` specifically from the `my-index-000001` index and `id4` from the `my-index-000002` index when the `user.query` matches the value `beagles`.

Reviewer 3: ```suggestion
* The second rule will generate a <<query-dsl-pinned-query>> pinning the <<mapping-id-field, `_id`>> of `id3` specifically from the `my-index-000001` index and `id4` from the `my-index-000002` index when the `user.query` metadata value matches `beagles`.
```

[source,console]
----
PUT /_query_rules/my-ruleset
{
  "rules": [
    {
      "rule_id": "rule1",
      "type": "pinned",
      "criteria": [
        {
          "type": "exact",
          "metadata": "user.query",
          "values": [ "pugs" ]
        }
      ],
      "actions": {
        "ids": [
          "id1",
          "id2"
        ]
      }
    },
    {
      "rule_id": "rule2",
      "type": "pinned",
      "criteria": [
        {
          "type": "exact",
          "metadata": "user.query",
          "values": [ "beagles" ]
        }
      ],
      "actions": {
        "docs": [
          {
            "_index": "my-index-000001",
            "_id": "id3"
          },
          {
            "_index": "my-index-000002",
            "_id": "id4"
          }
        ]
      }
    }
  ]
}
----

> // TEST[skip:TBD]

Reviewer 3: I guess tests will be added in a subsequent PR

Author: Good callout. Yes, I will plan a followup PR for tests.

The API response returns a results of `created` or `updated` depending on whether this was a new or edited ruleset.

[source,console-result]
----
{
  "result": "created"
}
----
// TEST[skip:TBD]

> You can use the <<get-query-ruleset>> call to retrieve the ruleset you just created,

Reviewer 3: Nice step-by-step instructions. I like how the page is like a tutorial 💯

the <<list-query-rulesets>> call to retrieve a summary of all query rulesets,
and the <<delete-query-ruleset>> call to delete a query ruleset.

[discrete]
[[rule-query-search]]
=== Perform a rule query

Once you have defined a query ruleset, you can search this ruleset using the <<query-dsl-rule-query>> query.
An example query for the `my-ruleset` defined above is:

[source,console]
----
GET /my-index-000001/_search
{
  "query": {
    "rule_query": {
      "organic": {
        "query_string": {
          "query": "pugs"
        }
      },

>       "match_criteria": {
>         "query_string": "pugs"
>       },

Reviewer 3: Would adding more match_criteria fields be confusing? I think it would help users understand the matching power of the query rules:

```suggestion
      "match_criteria": {
        "query_string": "pugs",
        "user_id": "kimchy"
      },
```

Author: Added!
